python/dgl/backend/jax/sparse.py

```diff
@@ -0,0 +1,79 @@
+import jax
+from jax import numpy as jnp
+from ...base import is_all, ALL
+from ...sparse import _gspmm, _gsddmm
+
+
+__all__ = ['gspmm', 'gsddmm', 'edge_softmax']
+
+
+def _reduce_grad(grad, shape):
+    """Reduce gradient on the broadcast dimension
+    If there is broadcast in forward pass, gradients need to be reduced on
+    broadcast dimension. This function checks the input tensor shape and
+    gradient shape and perform the reduction.
+
+    Parameters
+    ----------
+    grad: Tensor
+        Gradient tensor
+    shape: tuple
+        Shape of input tensor
+
+    Returns
+    -------
+    Tensor
+    """
+    grad_shape = grad.shape[1:]
+    in_shape = shape[1:]
+    if in_shape == grad_shape:
+        # no need to reduce
+        return grad
+    num_to_squeeze = len(grad_shape) - len(in_shape)
+    # pad inshape
+    in_shape = (1,) * num_to_squeeze + in_shape
+    reduce_idx = jnp.nonzero(jnp.tensor(grad_shape) - jnp.tensor(in_shape))
+    reduce_idx += 1  # skip batch dim
+    if len(reduce_idx) > 0:
+        grad = grad.sum(dim=tuple(reduce_idx), keepdim=True)
+    return grad.view(-1, *shape[1:])
+
+
+def _need_reduce_last_dim(ufeat, efeat):
+    """Indicates whether to reduce the last dimension on edges
+    in the backward pass of spmm,
+    if so, use dot instead of mul."""
+    ushp = ufeat.shape
+    eshp = efeat.shape
+    return ushp[1:-1] == eshp[1:-1] and eshp[-1] == 1 and ushp[-1] > 1
+
+
+def _muldiv(op, x):
+    return 1. / x if op == 'div' else x
+
+
+def _addsub(op, x):
+    return -x if op == 'sub' else x
+
+
+def _expand(x, shape):
+    return x.expand(-1, *shape)
+
+def gspmm(gidx, op, reduce_op, X, Y):
+    out, (argX, argY) = _gspmm(gidx, op, reduce_op, X, Y)
```

Comment:
Not that `_gspmm` and `_gsddmm` do not handle gradients.
In PyTorch backend we use `torch.autograd.Function` to manipulate gradients: https://github.com/dmlc/dgl/blob/master/python/dgl/backend/pytorch/sparse.py#L70-L118

Reply:
Yes! I was thinking about making sure all the forward passes are working properly before worrying about the gradients. One thing that's a bit tricky is that if we use things like `jax.custom_vjp` we'll have to think about how to cache and reuse stuff from the forward pass like this one: https://github.com/dmlc/dgl/blob/da9f14dff38c9939958ab1126438b7be2c3c39fc/python/dgl/backend/pytorch/sparse.py#L65